### PR TITLE
Uniffi library mode [ci full]

### PR DESCRIPTION
**Pull request by bendk**
This branch is testing out the new UniFFI library mode system and the results seem pretty good:

- The `uniffi.toml` files get simpler (although I'm not totally sure about the swift removal, since there's still some more work to do)
- It facilitates the transition from UDL -> proc-macros
- I think it makes it easier to use different megazords for focus vs firefox

There's still some work TODO:
- [x] The swift code needs some more work.  I ran the parts of `taskcluster/scripts/build-and-test-swift.py` that I could and the results were identical.  But there are some parts that need an Mac box, for example `megazords/ios-rust/build-xcframework.sh`).  I'm running full CI now, it'll be interesting to see what errors come back.
- [x] We need to land a few PRs on UniFFI and maybe do a release.
- [x] The gradle dependencies aren't totally right.  I was trying to run `test` and it failed because it depends on the release build of the megazord.

This part is speculative, but I think this could be taken further in the future.  We could remove the wrapper code from each component making them closer to vanilla Rust crates, no need for the `android` and `swift` directories.  The android/swift parts could be all in a single place that ran UniFFI library mode for a megazord and generated bindings for all crates that it depended on.  Removing the wrapper code means refactoring the Glean metrics, removing a backward compatibility layer, and adding UniFFI features like async callbacks so that we can express the API we want to export without needing an extra wrapper.  That's a lot of work, but many of those things come with their own motivation.  I know @tarikeshaq has a plan for reworking the Glean code, maybe it could be compatible with this dream.  The other changes would be nice for our documentation story -- as long as there's a wrapper layer, our Rust documentation won't accurately describe the consumer API.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

**Comment by mhammond** [comment]
For interest I had a poke at seeing if this works with the current uniffi version and it does :) There's a trivial fix to sync-manager importing `DeviceType` twice, and running the tests locally on a Mac failed due to `publish.gradle` failing a find `libmegazord.so` (it's `.dylib`) - and I failed to quickly google the best way to handle that elegantly, but hard-coding `.dylib` gets local tests passing.

**Comment by bendk** [comment]
I'm pushing an updated version of this that:
  - Fixes the merge conflicts
  - Removes the commit that was testing if we can use proc-macros.  I'd rather switch over fully to proc-macros than just do one function.  Maybe we can test that out after merging this
  - Has a blind fix for the libmegazord filename issue, let's see if it passes CI.

**Comment by bendk** [comment]
I think this is getting pretty close to ready.  @jeddai Do the changes to [run_python_tests.sh](https://github.com/mozilla/application-services/pull/5592/files#diff-caeeb7f604b92f1acb365eb94492a921c3cc2eab0348b6dca677d8ef9b21f95b) look good to you?  I'm actually a bit jealous of the cirrus situation, you're the first megazord where we can generate all the bindings with 1 command.

**Comment by jeddai** [comment]
> Do the changes to run_python_tests.sh look good to you?

Yep! I checked the python test logs and they look good.

> I'm actually a bit jealous of the cirrus situation, you're the first megazord where we can generate all the bindings with 1 command.

Oh I didn't realize that, that's awesome

**Comment by bendk** [comment]
@mhammond what do you think about merging this one to main, maybe after 117 starts?

**Comment by mhammond** [comment]
Happy for this to land whenever you feel it's ready!

File: components/suggest/android/build.gradle
Comment (linabutler) [inline components/suggest/android/build.gradle:8]: Thanks so much for migrating Suggest over, too!

Does its `uniffi.toml` need to change, too, to remove `cdylib_name` (and the `external_packages`, it looks like? 👀)

**Review by linabutler — APPROVED** [review]
Woohooo! So excited that this will unblock experimenting with UniFFI macros, too!

File: components/suggest/android/build.gradle
Comment (bendk) [reply components/suggest/android/build.gradle:8]: That makes sense to me, I just made that update.  Is there anything I should do to test if it's still working properly?

**Review by tarikeshaq — CHANGES_REQUESTED** [review]
Looks like this ended up being a breaking change for firefox-ios because of a swift name conflict - otherwise looks good! Maybe we can try to land https://github.com/mozilla/application-services/pull/5704 and use that as a confidence builder as well?

File: taskcluster/scripts/build-and-test-swift.py
Comment (bendk) [reply taskcluster/scripts/build-and-test-swift.py:13]: I don't think so, the bindings should automatically be generated.  UniFFI "knows" which UDL files were used to build a dylib and will automatically find them.

**Comment by bendk** [comment]
> Maybe we can try to land https://github.com/mozilla/application-services/pull/5704 and use that as a confidence builder as well?

+1 to this, I think that will go a long way.

File: megazords/ios-rust/build-xcframework.sh
Comment (bendk) [reply megazords/ios-rust/build-xcframework.sh:172]: I'm going to open a PR to rename the logger type, that seems like the simplest solution for now.

**Pull request closed without merging**
